lib/galaxy/files/sources/posix.py

```diff
@@ -62,6 +62,16 @@ def realize_to(self, source_path, native_path, user_context=None):
         else:
             shutil.move(source_native_path, native_path)
 
+    def _write_from(self, target_path, native_path, user_context=None):
+        effective_root = self._effective_root(user_context)
+        target_native_path = self._to_native_path(target_path, user_context=user_context)
+        # TODO: enforce_symlink_security
+        target_native_path = os.path.normpath(target_native_path)
+        assert target_native_path.startswith(os.path.normpath(effective_root))
+
+        # TODO: ensure directory exists, etc...
```

Comment:
@jmchilton this looks crucial.

Reply:
Yeah - I guess I need to write some more tests also. Thanks for catching that - I'll move this back into WIP.

Reply:
I think this is taken care of now? Plus some tests for link security and such.